docs/proposals/storage-capacity-proposal.md

```diff
@@ -0,0 +1,132 @@
+env:
+- name: NAMESPACE
+  valueFrom:
+    fieldRef:
+      fieldPath: metadata.namespace
+- name: POD_NAME
+  valueFrom:
+    fieldRef:
+      fieldPath: metadata.name
+```
+- Pass `--enable-capacity=true` to external-provisioner.
+- To enable ownership for CSIStorageCapacity objects and guarantee deleting objects when the driver is uninstalled, `--capacity-ownerref-level=2` argument must be passed to external-provisioner. The level indicates the number of objects that need to be traversed starting from the pod to reach the owning object for CSIStorageCapacity objects, which is `Deployment` now.
+- Add `storage.k8s.io` RBAC group
+- Following RBAC rules must be added to controller-rbac
+```
+  # Permissions for CSIStorageCapacity are only needed enabling the publishing
+  # of storage capacity information.
+  - apiGroups: ["storage.k8s.io"]
+    resources: ["csistoragecapacities"]
+    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
+  # The GET permissions below are needed for walking up the ownership chain
+  # for CSIStorageCapacity. They are sufficient for deployment via
+  # StatefulSet (only needs to get Pod) and Deployment (needs to get
+  # Pod and then ReplicaSet to find the Deployment).
+  - apiGroups: [""]
+    resources: ["pods"]
+    verbs: ["get"]
+  - apiGroups: ["apps"]
+    resources: ["replicasets"]
+    verbs: ["get"]
+```
+To enable using these object during scheduling `storageCapacity: true` must be added to CSIDriver spec.
+
+To implement `GetCapacity` it is possible to use AC. `GetCapacity` types look like this
+
+```
+
+type GetCapacityRequest struct {
+  // These are the same
+  // `volume_capabilities` the CO will use in `CreateVolumeRequest`.
+  VolumeCapabilities []*VolumeCapability
+  // These are the same `parameters` the CO will
+  // use in `CreateVolumeRequest`.
+  Parameters map[string]string
+  // This is the same as the
+  // `accessible_topology` the CO returns in a `CreateVolumeResponse`.
+  AccessibleTopology   *Topology
+}
+
+type GetCapacityResponse struct {
+  // The available capacity, in bytes, of the storage that can be used
+  AvailableCapacity int64
+  // The largest size that may be used in a
+  // CreateVolumeRequest.capacity_range.required_bytes field
+  // to create a volume with the same parameters as those in
+  // GetCapacityRequest.
+  MaximumVolumeSize *wrappers.Int64Value
+  // The smallest size that may be used in a
+  // CreateVolumeRequest.capacity_range.limit_bytes field
+  // to create a volume with the same parameters as those in
+  // GetCapacityRequest.
+  MinimumVolumeSize    *wrappers.Int64Value
+}
+```
+
+For each segment and each storage class, CSI GetCapacity is called once with the topology of the segment and the parameters of the class. For each request we should list all
+AC in cluster, filter out AC that do not satisfy segment (node), parameters and volume capacilities of request and find total size of remaining AC for `AvailableCapacity`, max AC
```

Comment:
How will that work for LVM storage classes? Will you count non LVG AC? Do we need to remove AC which are reserved when working together with scheduler extender?

Reply:
> How will that work for LVM storage classes?
For LVG SC Volumes capacityplanner now tries to select LVG AC and, in case they have been already reserved or their capacity is to small, tries non-LVG AC. So I suppose that we need to count non-LVG AC, because the volume may be placed on them too. Then MaximumVolumeSize would be max size from (LVG AC + non-LVG AC) and MinimumVolumeSize would be some constant threshold.

There I have a question. You use AcSizeMinThresholdBytes (1MB) as the minimum size of AC after which it must be deleted, but minimum Volume size for LVG is determined by DefaultPESize and equals 4MB. Which is the minimum LVG Volume size for plugin?

Reply:
> Do we need to remove AC which are reserved when working together with scheduler extender?

I suppose we should filter out reserved AC in case of non-LVG SC and subtract total reservation size from AC in case of LVG AC. There is no easy way to check reservations for AC, so we should list all ACR from cluster and loop over them, or add information about reservations in AC CRD

Reply:
@mishoyama I think that we should not consider reservations at all. The goal of proposed changes is to reduce number of nodes in filter() request (which is not obvious from proposal though). Also, we can't guarantee correct pod placement without ACR, so no changes required in this mechanism. For that goal, it is better if wrong nodes will be included in request, than if right nodes will be filtered out. For example, if we have 2 nodes with 1 drive on each, which are reserved for some pending pod, both nodes must be considered by extender for second pod, because it is actually feasible, we just need to wait until first pod will be placed and ACR will be deleted. Same logic is applied to LVM SC, we must count all AC, because only primary rough filtering is required from default scheduler, the main work does the extender